File: assertion/c/question3.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

_REPLACE_ME_WITH_DIRECTIVES_

int isSameNumber(int a, int b);

_REPLACE_ME_WITH_SOLUTION_

typedef struct TestCase
{
    int expected;
    int got;
    char *arguments;
} TestCase;

// creates a random number between min and max
int __randomNumber(int min, int max)
{
    return (rand() % (max - min + 1)) + min;
}

int main()
{
    srand(time(0));

    TestCase testCases[10] = {
        {.expected = 0,
         .got = isSameNumber(100, 212),
         .arguments = "isSameNumber(100, 212)"},
        {.expected = 1,
         .got = isSameNumber(25, 25),
         .arguments = "isSameNumber(25, 25)"}};
    ;

    for (int i = 2; i < 6; i++)
    {
        TestCase *test = testCases + i;

        int a = __randomNumber(0, 9999);
        int b = __randomNumber(0, 9999);
        int expected = a == b;
        int got = isSameNumber(a, b);
        test->got = got;
        test->expected = expected;
        test->arguments = malloc(sizeof(char[100]));
        sprintf(test->arguments, "isSameNumber(%d, %d)", a, b);
    }

    for (int i = 6; i < 10; i++)
    {
        TestCase *test = testCases + i;

        int a = __randomNumber(0, 9999);
        int expected = 1;
        int got = isSameNumber(a, a);
        test->got = got;
        test->expected = expected;
        test->arguments = malloc(sizeof(char[100]));
        sprintf(test->arguments, "isSameNumber(%d, %d)", a, a);
    }

    // `sizeof` returns the size of the memory used, not the length of the
    // array so we need to divide it by the size of the struct
    for (unsigned int i = 0; i < sizeof(testCases) / sizeof(TestCase); i++)
    {
        TestCase *test = testCases + i;

        if (test->got == test->expected)
        {
            printf("# %d PASSING\n", i + 1);
        }
        else
        {
            printf("# %d FAILED\n", i + 1);
        }

        printf("> ARGUMENTS %s\n", test->arguments);
        printf("> EXPECTED %d\n", test->expected);
        printf("> GOT %d\n", test->got);

        if (i >= 2) {
            free(test->arguments);
        }
    }
    return 0;
}

File: assertion/php/question3.php
<?php

_REPLACE_ME_WITH_DIRECTIVES_

_REPLACE_ME_WITH_SOLUTION_

function main(): void {
    $test_cases = [
        [
            "arguments" => "isSameNumber(100, 212)",
            "got" => isSameNumber(100, 212),
            "expected" => false
        ],
        [
            "arguments" => "isSameNumber(25, 25)",
            "got" => isSameNumber(25, 25),
            "expected" => true
        ],
    ];

    for ($i = 0; $i < 4; $i++) {
        $a = rand(0, 9999);
        $b = rand(0, 9999);
        $expected = $a === $b;
        $got = isSameNumber($a, $b);
        $arguments = "isSameNumber($a, $b)";
        array_push($test_cases, [
            "got" => $got,
            "expected" => $expected,
            "arguments" => $arguments
        ]);
    }

    for ($i = 0; $i < 4; $i++) {
        $a = rand(0, 9999);
        $expected = true;
        $got = isSameNumber($a, $a);
        $arguments = "isSameNumber($a, $a)";
        array_push($test_cases, [
            "got" => $got,
            "expected" => $expected,
            "arguments" => $arguments
        ]);
    }

    for ($i = 0; $i < count($test_cases); $i++) {
        $test = $test_cases[$i];

        if ($test["got"] === $test["expected"]) {
            echo "# ". $i + 1 . " PASSING\n";
            echo "> ARGUMENTS " . $test["arguments"] . "\n";
            echo "> EXPECTED " . $test["expected"] . "\n";
            echo "> GOT " . $test["got"] . "\n";
        } else {
            echo "# " . $i + 1 . " FAILED\n";
            echo "> ARGUMENTS " . $test["arguments"] . "\n";
            echo "> EXPECTED " . $test["expected"] . "\n";
            echo "> GOT " . $test["got"] . "\n";
        }
    }
}

main();